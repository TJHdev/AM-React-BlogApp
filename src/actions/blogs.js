import database from '../firebase/firebase.js'

// component calls action generator
// action generator returns object
// component dispatches object
// redux store changes

// component calls action generator
// action generator returns function
// component dispatches object (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)

export const addBlog = (blog) => ({
  type: 'ADD_BLOG',
  blog: blog
})

export const startAddBlog = (blogData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      title = '',
      body = '',
      createdAt = 0,
      lastEditedAt = 0
    } = blogData;
    
    const blog = { title, body, createdAt, lastEditedAt };

    return database.ref(`users/${uid}/blogs`).push(blog).then((ref) => {
      dispatch(addBlog({
        id: ref.key,
        ...blog
      }));
    });
  };
};

export const removeBlog = ({ id } = {}) => ({
  type: 'REMOVE_BLOG',
  id: id
});

export const startRemoveBlog = ({ id }) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).set(null).then(() => {
      dispatch(removeBlog({ id }));
    });
  }
};

export const editBlog = (id, updates) => ({
  type: 'EDIT_BLOG',
  id: id,
  updates
});

export const startEditBlog = (id, updates = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs/${id}`).update(updates).then(() => {
      dispatch(editBlog(id, updates));
    }); 
  };
};

export const setBlogs = (blogs) => ({
  type: 'SET_BLOGS',
  blogs: blogs
});

export const startSetBlogs = (blogsData = []) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/blogs`).once('value').then((snapshot) => {
      const blogs = [];
      snapshot.forEach((childSnapshot) => {
        blogs.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

export const startSetBlogsPublic = (blogsData = []) => {
  return (dispatch, getState) => {
    return database.ref('users').once('value').then((snapshot) => {
      const blogs = [];
      const val = snapshot.val();

      Object.keys(val).forEach((childSnapshot) => {
        Object.keys(val[childSnapshot].blogs).forEach((blog) => {
          blogs.push({
            id: blog,
            ...snapshot.val()[childSnapshot].blogs[blog]
          })
        })
      })
      dispatch(setBlogs(blogs));
    });
  }
};

