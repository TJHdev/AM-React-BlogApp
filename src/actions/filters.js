import moment from 'moment';

const filtersReducerDefaultState = {
  text: '',
  sortBy: 'date',
};

const filtersReducer = (state = filtersReducerDefaultState, action) => {
  switch (action.type) {
    case 'SET_TEXT_FILTER':
      return {
        ...state,
        text: action.text
      };
    case 'SORT_BY_DATE':
      return {
        ...state,
        sortBy: 'date'
      };
    case 'SORT_BY_TITLE':
      return {
        ...state,
        sortBy: 'title'
      };
    default: 
      return state;
  }
};

export default filtersReducer;