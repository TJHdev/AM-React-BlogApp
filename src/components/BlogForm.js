import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

const now = moment();

export default class BlogForm extends React.Component {
  constructor(props) {
    super(props);
    
    this.state = { // no way to do add/edit conditional without constructor
      title: props.blog ? props.blog.title : '',
      body: props.blog ? props.blog.body :  '',
      createdAt: props.blog ? moment(props.blog.createdAt) :  moment(),
      lastEditedAt: props.blog ? moment(props.blog.lastEditedAt) :  moment(),
      calendarFocused: false,
      error: ''
    };
  };
  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };
  onBodyChange = (e) => {
    const body = e.target.value;
    this.setState(() => ({ body }));
    // // alternetive implementation
    // e.persist();
    // this.setState(() => ({ e.target.value }));
  };
  onDateChange = (createdAt) => {
    if (createdAt) { // prevents the user from clearing this value
      this.setState(() => ({ createdAt }));
    }
  };
  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.title || !this.state.body) {
      this.setState(() => ({ error: 'Please provide title and body!' }))
    } else {
      this.setState(() => ({ error: '' })) 
      this.props.onSubmit({
        title: this.state.title,
        body: this.state.body,
        createdAt: this.state.createdAt.valueOf(),
        lastEditedAt: moment().valueOf()
      }) // we pass the data up so that the form can remain re-usable
    }
  };
  render() {
    return (
        <form className="form" onSubmit={this.onSubmit}>
          {this.state.error && <p className="form__error">{this.state.error}</p>}
          <input
            type="text"
            placeholder="Title"
            className="text-input"
            autoFocus
            value={this.state.title}
            onChange={this.onTitleChange}
          />
          <textarea
            placeholder="Add the text body"
            className="textarea"
            value={this.state.body}
            onChange={this.onBodyChange}
          ></textarea>
          <div>
            <button className="button">Save Blog</button>
          </div>
        </form>
    );
  }
}