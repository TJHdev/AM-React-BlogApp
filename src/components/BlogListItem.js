import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const BlogListItem = ({ id, title, createdAt }) => (
    <Link className="list-item" to={`/edit/${id}`}>
      <h3 className="list-item__title">{title}</h3>
      <span className="list-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</span>
    </Link>
);

export default BlogListItem;