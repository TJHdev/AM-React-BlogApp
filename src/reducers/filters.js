import moment from 'moment';

const filtersReducerDefaultState = {
  text: '',
  sortBy: 'dateCreated',
};

const filtersReducer = (state = filtersReducerDefaultState, action) => {
  switch (action.type) {
    case 'SET_TEXT_FILTER':
      return {
        ...state,
        text: action.text
      };
    case 'SORT_BY_DATE_CREATED':
      return {
        ...state,
        sortBy: 'dateCreated'
      };
    case 'SORT_BY_DATE_EDITED':
      return {
        ...state,
        sortBy: 'dateEdited'
      };
    case 'SORT_BY_TITLE':
      return {
        ...state,
        sortBy: 'title'
      };
    default: 
      return state;
  }
};

export default filtersReducer;