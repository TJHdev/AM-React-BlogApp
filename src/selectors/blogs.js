import moment from 'moment';

// Get visible blogs

const getVisibleBlogs = (blogs, { text, sortBy }) => {
  return blogs.filter((blog) => {
    const textMatch = blog.title.toLowerCase().includes(text.toLowerCase());
    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'dateCreated') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'dateEdited') {
      return a.lastEditedAt < b.lastEditedAt ? 1 : -1;
    } else if (sortBy === 'title') {
      return a.title < b.title ? 1 : -1;
    }
  });
};

export default getVisibleBlogs;