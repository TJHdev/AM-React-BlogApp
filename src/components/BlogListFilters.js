import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { setTextFilter, sortByDateCreated, sortByDateEdited, sortByTitle } from '../actions/filters';

export class BlogListFilters extends React.Component { // this allows us to add state
  state = {
    calendarFocused: null
  };
  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };
  onSortChange = (e) => {
    if (e.target.value === 'dateCreated') {
      this.props.sortByDateCreated();
    } else if (e.target.value === 'dateEdited') {
      this.props.sortByDateEdited();
    } else if (e.target.value === 'title') {
      this.props.sortByTitle();
    }
  }
  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input 
              className="text-input"
              type="text"
              autoFocus
              defaultValue={this.props.filters.text}
              onChange={this.onTextChange}
              placeholder={'Search blogs'} 
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="dateCreated">Date Created</option>
              <option value="dateEdited">Date Last Edited</option>
              <option value="title">Title</option>
            </select>
          </div>
          { this.props.isAuthenticated && 
            <div className="input-group__item">
              <Link className="button" to="/create">Add Blog</Link>
            </div>
          }
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
    isAuthenticated: Boolean(state.auth.uid)
  };
};

const mapDispatchToProps = (dispatch) => (
  {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDateCreated: () => dispatch(sortByDateCreated()),
    sortByDateEdited: () => dispatch(sortByDateEdited()),
    sortByTitle: () => dispatch(sortByTitle()),
  }
)

export default connect(mapStateToProps, mapDispatchToProps)(BlogListFilters);