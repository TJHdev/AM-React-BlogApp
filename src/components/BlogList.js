import React from 'react';
import { connect } from 'react-redux';
import BlogListItem from './BlogListItem';
import selectBlogs from '../selectors/blogs';

export const BlogList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Blogs</div>
      <div className="show-for-desktop">Blog</div>
      <div className="show-for-desktop">Date</div>
    </div>
    <div className="list-body">
      {
        props.blogs.length === 0 ? (
          <div className="list-item list-item--message">
            <span>There are no blogs</span>
          </div>
        ) : (
          props.blogs.map((blog) => (
            <BlogListItem key={blog.id} {...blog} />
          ))
        )
      }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    blogs: selectBlogs(state.blogs, state.filters)
  };
};

export default connect(mapStateToProps)(BlogList);
