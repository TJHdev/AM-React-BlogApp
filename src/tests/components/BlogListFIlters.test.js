import React from 'react';
import { shallow } from 'enzyme';
import blogs from '../fixtures/blogs';
import { filters, altFilters } from '../fixtures/filters';
import { BlogListFilters } from '../../components/BlogListFilters';
import moment from 'moment';

let setTextFilter, sortByDateCreated, sortByDateEdited, sortByTitle, wrapper;

beforeEach(() => {
  setTextFilter = jest.fn();
  sortByDateCreated = jest.fn();
  sortByDateEdited = jest.fn();
  sortByTitle = jest.fn();
  wrapper = shallow(
    <BlogListFilters 
      filters={filters}
      setTextFilter={setTextFilter}
      sortByDateCreated={sortByDateCreated}
      sortByDateEdited={sortByDateEdited}
      sortByTitle={sortByTitle}
    />
  );
});


test('should render BlogListFilters correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should render BlogListFilters with alt data correctly', () => {
  wrapper.setProps({
    filters: altFilters
  });
  expect(wrapper).toMatchSnapshot();
});

test('should handle text change', () => {
  const value = 'mech';
  wrapper.find('input').prop('onChange')({target: {value}});
  expect(setTextFilter).toHaveBeenCalledWith('mech');
});

test('should sort by date created', () => {
  wrapper.setProps({
    filters: altFilters
  })
  const value = 'dateCreated';
  wrapper.find('select').prop('onChange')({target: {value}});
  expect(sortByDateCreated).toHaveBeenCalledTimes(1);

});

test('should sort by date last edited', () => {
  const value = 'dateEdited';
  wrapper.find('select').prop('onChange')({target: {value}});
  expect(sortByDateEdited).toHaveBeenCalledTimes(1);
});


test('should sort by title', () => {
  const value = 'title';
  wrapper.find('select').prop('onChange')({target: {value}});
  expect(sortByTitle).toHaveBeenCalledTimes(1);
});