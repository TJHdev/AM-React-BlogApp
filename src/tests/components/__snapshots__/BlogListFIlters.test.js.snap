// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render BlogListFilters correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        autoFocus={true}
        className="text-input"
        defaultValue=""
        onChange={[Function]}
        placeholder="Search blogs"
        type="text"
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
        value="dateCreated"
      >
        <option
          value="dateCreated"
        >
          Date Created
        </option>
        <option
          value="dateEdited"
        >
          Date Last Edited
        </option>
        <option
          value="title"
        >
          Title
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`should render BlogListFilters with alt data correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="input-group"
  >
    <div
      className="input-group__item"
    >
      <input
        autoFocus={true}
        className="text-input"
        defaultValue="mech"
        onChange={[Function]}
        placeholder="Search blogs"
        type="text"
      />
    </div>
    <div
      className="input-group__item"
    >
      <select
        className="select"
        onChange={[Function]}
        value="title"
      >
        <option
          value="dateCreated"
        >
          Date Created
        </option>
        <option
          value="dateEdited"
        >
          Date Last Edited
        </option>
        <option
          value="title"
        >
          Title
        </option>
      </select>
    </div>
  </div>
</div>
`;
